### PR TITLE
add support for local file ignore_extension

### PR DESCRIPTION
**Pull request by shmuelamar**
fix for #172 , few notes:
1. Ill will add tests if all the rest looks fine
2. i used the `ClosingGzipFile` class for S3 instead of plain `GzipFile`. reviewing the python3 code for `GzipFile` it seems to not affect but its something worth reviewing to see if it affects something (it even might be better). LMK if you want me to use for S3 the plain stdlib `GzipFile` and ill change the implementation.
3. it will be nice to add `ignore_extension` to docs, to find it in the first place I dig into the source.

### CODE REVIEW

File: smart_open/smart_open_lib.py [hunk 1]
```diff
@@ -586,24 +552,29 @@ def close(self):
...
     file_obj.
     """
+    if ignore_extension:
+        return file_obj
+
     _, ext = os.path.splitext(filename)
     if ext == '.bz2':
         return ClosingBZ2File(file_obj, mode)
-    elif ext == '.gz':
+    if ext == '.gz':
```
Comment (menshikh-iv) [inline smart_open/smart_open_lib.py]: why this change?

File: smart_open/tests/test_smart_open.py [hunk 2]
```diff
@@ -292,29 +292,29 @@ def test_file(self, mock_smart_open):
         smart_open_object = smart_open.smart_open(prefix+full_path, read_mode)
         smart_open_object.__iter__()
         # called with the correct path?
-        mock_smart_open.assert_called_with(full_path, read_mode, encoding=None, errors='strict')
+        mock_smart_open.assert_called_with(full_path, read_mode, encoding=None, errors='strict', ignore_extension=False)
```
Comment (menshikh-iv) [inline smart_open/tests/test_smart_open.py]: Need also add a test for check this feature (where `ignore_extension=True`)

**Comment by menshikh-iv** [comment]
ping @shmuelamar, are you plan to finish this PR?

**Comment by shmuelamar** [comment]
yes @menshikh-iv ill try finish it this week - got busy week.

**Comment by shmuelamar** [comment]
@menshikh-iv just finished completing the tests and reverting the if/else change.
also added a README example.

File: smart_open/smart_open_lib.py [hunk 3]
```diff
@@ -215,7 +219,7 @@ def smart_open(uri, mode="rb", **kw):
         raise TypeError('don\'t know how to handle uri %s' % repr(uri))
 
 
-def s3_open_uri(parsed_uri, mode, **kwargs):
+def s3_open_uri(parsed_uri, mode, ignore_extension=False, **kwargs):
```
Comment (menshikh-iv) [inline smart_open/smart_open_lib.py]: Should `ignore_extension` be an explicit parameter (or this should be part of `kwargs`), same question for `s3_open_key`, @mpenkov thought?

File: smart_open/tests/test_smart_open.py [hunk 4]
```diff
@@ -292,29 +319,40 @@ def test_file(self, mock_smart_open):
...
         full_path = os.path.expanduser(short_path)
 
-        smart_open_object = smart_open.smart_open(prefix+short_path, read_mode, errors='strict')
+        smart_open_object = smart_open.smart_open(prefix+short_path, read_mode, errors='strict', ignore_extension=False)
         smart_open_object.__iter__()
         # called with the correct expanded path?
-        mock_smart_open.assert_called_with(full_path, read_mode, encoding=None, errors='strict')
+        mock_smart_open.assert_called_with(full_path, read_mode, encoding=None, errors='strict', ignore_extension=False)
+
+    @mock.patch('smart_open.smart_open_lib.file_smart_open')
```
Comment (menshikh-iv) [inline smart_open/tests/test_smart_open.py]: should it be `maybe_..` decorator instead?

**Comment by shmuelamar** [comment]
hey @menshikh-iv @mpenkov pls LMK how can I help merge this PR. is there anything else to fix or answer?

**Comment by menshikh-iv** [comment]
@shmuelamar please resolve merge-conflict.
@mpenkov please have a look.

**Comment by menshikh-iv** [comment]
@shmuelamar we'll merge #185 first and after - current PR, don't worry.

**Comment by menshikh-iv** [comment]
@shmuelamar can you make rebase with current master, please?

**Comment by mpenkov** [comment]
The current master handles ignore_extension across all interfaces. It's now in one place: https://github.com/RaRe-Technologies/smart_open/blob/master/smart_open/smart_open_lib.py#L195 

@shmuelamar please have a look there, it should help with your rebase.

**Comment by menshikh-iv** [comment]
ping @shmuelamar, what's about PR?

**Comment by menshikh-iv** [comment]
Ping @shmuelamar, are you planning to finish PR?

**Comment by mpenkov** [comment]
Abandoned.

Furthermore, the current code seems to include the feature mentioned in this PR.

```code
$ python -c 'import smart_open as s;print(s.smart_open("test.txt.gz").read())'
b'hello world\n'
$ python -c 'import smart_open as s;print(s.smart_open("test.txt.gz", ignore_extension=True).read())'
b'\x1f\x8b\x08\x08\xd8\xdbt\\\x00\x03test.txt\x00\xcbH\xcd\xc9\xc9W(\xcf/\xcaI\xe1\x02\x00-;\x08\xaf\x0c\x00\x00\x00'
$
```
